### ISSUE TITLE
`shared_sections` in libia2 should be left unprotected

### ISSUE THREAD

**Issue by endbr64** [immunant/IA2-Phase2#259]
This static is needed for subsequent compartment initialization. There is probably more data like this that should be shared.

**Comment by fw-immunant** [comment]
What does it take to reproduce a failure here? Is it something that only shows up with >2 compartments?

**Comment by endbr64** [comment]
Yep, after the main binary is protected, `shared_sections` becomes inaccessible. Right now we protect the main binary last, but that could be a coincidence. When the Perl module is loaded later, it crashes.

Just sharing the pointer is enough to make it work again. https://github.com/immunant/IA2-Phase2/commit/1b3f9ba05aa52ac5feb739f7961d6975741e0153

**Comment by rinon** [comment]
I wonder why this variable is being put in a writable segment, it should be read-only.

**Comment by fw-immunant** [comment]
> I wonder why this variable is being put in a writable segment, it should be read-only.

```
$ cat test.c
static const char *shared_sections[][2] = {
    {"__start_ia2_shared_data", "__stop_ia2_shared_data"},
};

static const char *const shared_sections2[][2] = {
    {"__start_ia2_shared_data", "__stop_ia2_shared_data"},
};

int main(int argc, char** argv) {
    return 0;
}
$ gcc test.c
$ objdump --syms a.out | grep shared_sections
0000000000004010 l     O .data	0000000000000010              shared_sections
0000000000003e00 l     O .data.rel.ro	0000000000000010              shared_sections2
```

**Comment by rinon** [comment]
C pointer syntax strikes again. I'll fix this.

**Comment by rinon** [comment]
Fixed in 5a85caebfa319b077abf820c80ead76f66e3f63d
